### PR TITLE
[#159730005] Generate bosh runtime-config variables in the bosh update-runtime-config task

### PR DESCRIPTION
**Pull request by tlwr**
tldr;

- do not copy cf-vars-store anymore
- do not have a separate task for bosh-runtime-config interpolation, do it in bosh update-runtime config step instead.

---

# What

Previously we had a task `steal-cf-vars` which copied bosh-dns related
pki material from `cf-vars-store` to `bosh-runtime-config-vars-store`

Now that bosh-dns has been rolled out everywhere (that we care about:
staging, prod) we can get rid of the copying task. This task is now
inlined into the runtime config task, which generates said variables if they do not exist.

Expected behaviour is that new environments have the vars generated, and
existing environments use the existing vars.

# How to review

- Code review

#### First

- Take a copy of `bosh-runtime-config-vars-store.yml`
- Run `create-bosh-concourse`
- Ensure that the copy of `bosh-runtime-config-vars-store.yml` is the same

#### Then

- Delete `bosh-runtime-config-vars-store.yml`
- Run `create-bosh-concourse`
- Green

### CODE REVIEW

File: concourse/pipelines/create-bosh-concourse.yml [hunk 1]
```diff
@@ -700,6 +694,38 @@ jobs:
...
+            args:
+              - -ec
+              - |
+                bosh interpolate \
+                  --var-errs \
+                  --vars-store bosh-runtime-config-vars-store/bosh-runtime-config-vars-store.yml \
+                  paas-bootstrap/manifests/runtime-config/variables-template.yml
+
+                cp bosh-runtime-config-vars-store/bosh-runtime-config-vars-store.yml \
+                   new-bosh-runtime-config-vars-store/bosh-runtime-config-vars-store.yml
```
Comment (alext) [inline concourse/pipelines/create-bosh-concourse.yml]: Is there a reason this needs to be a separate task like this. Could it not be part of the set-runtime-configs task? It would make sense to me to have all the needed variables defined in the runtime-config manifest, and get rid of the `variables-template.yml` file. This would be more consistent with other manifests (eg the CF manifest).

**Comment by tlwr** [comment]
Addressed the above comments, updated the PR description.

**Review by jpluscplusm — APPROVED** [review]
LGTM.

~~Suggestions:~~

~~- include `BOSH_NON_INTERACTIVE` in the remaining task's envvars.~~

**Review by richardTowers — APPROVED** [review]

File: concourse/pipelines/create-bosh-concourse.yml [hunk 2]
```diff
@@ -1042,10 +1007,17 @@ jobs:
                   --name '' \
                   blank-runtime-config.yml
 
+                cp bosh-runtime-config-vars-store/bosh-runtime-config-vars-store.yml \
+                   new-bosh-runtime-config-vars-store/bosh-runtime-config-vars-store.yml
+
                 bosh update-runtime-config \
```
Comment (keymon) [inline concourse/pipelines/create-bosh-concourse.yml]: I thik it is good to still do two steps calling `bosh interpolate` and then `bosh update-runtime-config`.

The reason is that in `bosh interpolate` you can pass this two handy parameters:

```
          --var-errs             Expect all variables to be found, otherwise error
          --var-errs-unused      Expect all variables to be used, otherwise error
```

File: manifests/runtime-config/bosh-dns.yml [hunk 3]
```diff
@@ -35,3 +35,62 @@ addons:
...
+
+- name: dns_api_client_tls
+  type: certificate
+  options:
+    ca: dns_api_tls_ca
+    common_name: api.bosh-dns
+    extended_key_usage:
+    - client_auth
+
+## Our additions
```
Comment (keymon) [inline manifests/runtime-config/bosh-dns.yml]: I wouldn't put that comment. It would get forgotten :)

File: manifests/runtime-config/bosh-dns.yml [hunk 4]
```diff
@@ -35,3 +35,62 @@ addons:
...
+
+- name: dns_healthcheck_tls_ca_old
```
Comment (keymon) [inline manifests/runtime-config/bosh-dns.yml]: BTW, I am not able to find where we implement the certificate rotation for this project. We do in paas-cf, but not sure if we do here. 

Should we include these variables if they are not useful?

File: concourse/pipelines/create-bosh-concourse.yml (hunk 2 above)
Comment (jpluscplusm) [reply concourse/pipelines/create-bosh-concourse.yml]: Thanks, @keymon - I *knew* that `update-runtime-config` doesn't accept `--vars-errs` and that was why I'd broken it into 2 steps (or, at least, 2 commands), but I didn't flag it up in review yesterday.  Good call.

File: manifests/runtime-config/bosh-dns.yml (hunk 4 above)
Comment (jpluscplusm) [reply manifests/runtime-config/bosh-dns.yml]: These vars are needed to satisfy the static analysis tests in Travis, which insist on all `*ca*` things having an `_old` variant concatenated.  We could get by with dummy values in there, not backed by bosh-generated "proper" CAs, but it seemed like more work to do that than simply to include them as carbon copies of their non-`_old` siblings.

Comment (keymon) [reply manifests/runtime-config/bosh-dns.yml]: Oh, do we have such test in this project?

File: concourse/pipelines/create-bosh-concourse.yml (hunk 2 above)
Comment (jpluscplusm) [reply concourse/pipelines/create-bosh-concourse.yml]: @keymon Addressed in the recent push.

File: manifests/runtime-config/bosh-dns.yml (hunk 4 above)
Comment (jpluscplusm) [reply manifests/runtime-config/bosh-dns.yml]: I really thought we did, but it's very possible we'd assumed such a test existed, based on paas-cf.

Given the `_old` vars are in production right now, I don't think the overhead of having them there is significant - it would be more effort to remove them.

I don't feel this particularly /strongly/, however :-) @alext @richardTowers @henrytk opinions?

Comment (keymon) [reply manifests/runtime-config/bosh-dns.yml]: IMO I would delete the `_old` if it is not being tested and the feature to rotate does not exist. It is misleading and it does not provide anything.

Also it is easy to remove and to add later if needed.

Comment (richardTowers) [reply manifests/runtime-config/bosh-dns.yml]: I think let's remove them.

**Review by keymon — APPROVED** [review]
Despite the `_old` cert, this looks good for me.

**Comment by jpluscplusm** [comment]
Addressed all comments.

File: manifests/runtime-config/bosh-dns.yml (hunk 4 above)
Comment (jpluscplusm) [reply manifests/runtime-config/bosh-dns.yml]: @richardTowers @keymon Done. Please re-review.

Comment (keymon) [reply manifests/runtime-config/bosh-dns.yml]: LGTM. I am rerunning in my env and would merge soon.

**Comment by keymon** [comment]
> You can retrieve the config from the bosh director with bosh config runtime --name=bosh-dns

Yes, as you say that is the **current** config, but we do not store a history of the configurations. 

I think that now we are adding the bosh-dns here, this is a critical piece of config and we should be able to audit it if needed. We should know why a changed was triggered and what changed.

**Comment by jpluscplusm** [comment]
@keymon 
> history of the configurations

That's "this repo", isn't it? Apart from pushing a version containing the certs, which feels a bit odd from a secrets perspective, what would storing the deployed manifest add to our view of history?
